src/plugins/embeddable/public/lib/state_transfer/embeddable_state_transfer.ts

```diff
@@ -0,0 +1,120 @@
+ */
+export class EmbeddableStateTransfer {
+  constructor(
+    private navigateToApp: ApplicationStart['navigateToApp'],
+    private scopedHistory?: ScopedHistory
+  ) {}
+
+  /**
+   * Fetches an {@link EmbeddableOriginatingAppState | originating app} argument from the scoped
+   * history's location state.
+   *
+   * @param history - the scoped history to fetch from
+   * @param removeAfterFetch - determines whether or not the state transfer service removes all keys relating to this object
+   */
+  public getIncomingOriginatingApp(
+    removeAfterFetch: boolean = false
+  ): EmbeddableOriginatingAppState | undefined {
+    return this.getIncomingState<EmbeddableOriginatingAppState>(
+      isEmbeddableOriginatingAppState,
+      removeAfterFetch
+    );
+  }
+
+  /**
+   * Fetches an {@link EmbeddablePackageState | embeddable package} argument from the scoped
+   * history's location state.
+   *
+   * @param history - the scoped history to fetch from
+   * @param removeAfterFetch - determines whether or not the state transfer service removes all keys relating to this object
+   */
+  public getIncomingEmbeddablePackage(
+    removeAfterFetch: boolean = false
+  ): EmbeddablePackageState | undefined {
+    return this.getIncomingState<EmbeddablePackageState>(
+      isEmbeddablePackageState,
+      removeAfterFetch
+    );
+  }
+
+  /**
+   * A wrapper around the {@link ApplicationStart.navigateToApp} method which navigates to the specified appId
+   * with {@link EmbeddableOriginatingAppState | originating app state}
+   */
+  public async navigateToWithOriginatingApp(
+    appId: string,
+    options: { path?: string; state: EmbeddableOriginatingAppState }
+  ): Promise<void> {
+    await this.navigateToWithState<EmbeddableOriginatingAppState>(appId, options);
+  }
+
+  /**
+   * A wrapper around the {@link ApplicationStart.navigateToApp} method which navigates to the specified appId
+   * with {@link EmbeddablePackageState | embeddable package state}
+   */
+  public async navigateToWithEmbeddablePackage(
+    appId: string,
+    options: { path?: string; state: EmbeddablePackageState }
+  ): Promise<void> {
+    await this.navigateToWithState<EmbeddablePackageState>(appId, options);
+  }
+
+  private getIncomingState<IncomingStateType>(
+    guard: (state: unknown) => state is IncomingStateType,
+    removeAfterFetch: boolean = false
+  ): IncomingStateType | undefined {
+    if (!this.scopedHistory) {
+      throw new TypeError('ScopedHistory is required to fetch incoming state');
+    }
+    const incomingState = this.scopedHistory.location?.state;
+    const castState =
+      !guard || guard(incomingState) ? (cloneDeep(incomingState) as IncomingStateType) : undefined;
+    if (castState && removeAfterFetch) {
+      Object.keys(castState).forEach((key: string) => {
+        delete (this.scopedHistory!.location.state as { [key: string]: unknown })[key];
+      });
+      this.scopedHistory.push(this.scopedHistory.location);
```

Comment:
I originally found that replace caused the location to be updated, and the application to reload, thus losing the information which was fetched from state. After a bit more research, it seems like `replace` is the right option, and I just needed to send back the same location to avoid a reload.